File: ultra-fast.controller.ts
import { Controller, Post, Body, Get, Res, Req } from '@nestjs/common';
import { FastifyReply, FastifyRequest } from 'fastify';

// Ultra-fast SQLite setup
let db: any;
let insertStmt: any;
let isBun = false;
let counter = 0;

// Initialize database on module load
if (typeof Bun !== 'undefined') {
    isBun = true;
    const { Database } = require('bun:sqlite');
    db = new Database(':memory:');
    
    // Create table
    db.exec(`CREATE TABLE iot_payload (
        id TEXT PRIMARY KEY,
        content TEXT,
        ts INTEGER
    )`);
    
    // Pre-prepare statement for maximum speed
    insertStmt = db.prepare('INSERT INTO iot_payload (id, content, ts) VALUES (?, ?, ?)');
    console.log('Ultra-Fast: Using Bun native SQLite');
} else {
    const sqlite3 = require('sqlite3').verbose();
    db = new sqlite3.Database(':memory:');
    
    db.serialize(() => {
        db.run(`CREATE TABLE iot_payload (
            id TEXT PRIMARY KEY,
            content TEXT,
            ts INTEGER
        )`);
    });
    console.log('Ultra-Fast: Using Node.js SQLite3');
}

// Pre-allocated objects to avoid GC pressure
const responseTemplate = { id: '', t_ms: 0 };
const healthTemplate = { status: 'ok', timestamp: '', runtime: '', count: 0 };

@Controller()
export class UltraFastController {

  @Post('ingest')
  async ingest(
    @Body() payload: any,
    @Req() request: FastifyRequest,
    @Res() reply: FastifyReply
  ) {
    const start = process.hrtime.bigint();
    
    try {
      // Ultra-fast ID generation
      const id = `id_${++counter}_${Date.now()}`;
      const content = JSON.stringify(payload);
      const timestamp = Date.now();
      
      // Ultra-fast database insert
      if (isBun) {
        // Bun native SQLite - should be ultra fast
        insertStmt.run(id, content, timestamp);
      } else {
        // Node.js SQLite3 - async but optimized
        await new Promise<void>((resolve, reject) => {
          db.run(
            'INSERT INTO iot_payload (id, content, ts) VALUES (?, ?, ?)',
            [id, content, timestamp],
            function(err: any) {
              if (err) reject(err);
              else resolve();
            }
          );
        });
      }
      
      // No background processing - just return immediately
      const end = process.hrtime.bigint();
      const elapsed = Number(end - start) / 1000000;
      
      // Reuse object to avoid allocation
      responseTemplate.id = id;
      responseTemplate.t_ms = Math.round(elapsed * 100) / 100;
      
      reply.send(responseTemplate);
      
    } catch (error) {
      console.error('Error:', error);
      reply.code(500).send({ error: 'Internal error' });
    }
  }

  @Get('health')
  health(@Res() reply: FastifyReply) {
    // Reuse object to avoid allocation
    healthTemplate.timestamp = new Date().toISOString();
    healthTemplate.runtime = typeof Bun !== 'undefined' ? `Bun ${Bun.version}` : `Node.js ${process.version}`;
    healthTemplate.count = counter;
    
    reply.send(healthTemplate);
  }
}

File: src/main/java/com/example/iotbench/UltraFastController.java
package com.example.iotbench;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.jdbc.core.JdbcTemplate;
import java.util.*;
import java.util.concurrent.atomic.AtomicLong;

@RestController
@RequestMapping("/ultra")
public class UltraFastController {

    private final JdbcTemplate jdbc;
    private final AtomicLong counter = new AtomicLong(0);
    
    // Pre-allocated objects to avoid GC pressure (like the NestJS version)
    private final Map<String, Object> responseTemplate = new HashMap<>();
    private final Map<String, Object> healthTemplate = new HashMap<>();

    public UltraFastController(JdbcTemplate jdbc) {
        this.jdbc = jdbc;
        initTable();
        
        // Pre-populate response templates
        responseTemplate.put("id", "");
        responseTemplate.put("t_ms", 0.0);
        
        healthTemplate.put("status", "ok");
        healthTemplate.put("timestamp", "");
        healthTemplate.put("runtime", "Spring Boot");
        healthTemplate.put("count", 0L);
    }

    private void initTable() {
        jdbc.execute("CREATE TABLE IF NOT EXISTS iot_payload_ultra (id VARCHAR(36) PRIMARY KEY, content CLOB, ts BIGINT)");
    }

    @PostMapping("ingest")
    public ResponseEntity<Map<String,Object>> ingest(@RequestBody Map<String,Object> payload) {
        long start = System.nanoTime();
        
        try {
            // Ultra-fast ID generation (like NestJS version)
            long count = counter.incrementAndGet();
            String id = "id_" + count + "_" + System.currentTimeMillis();
            
            // Convert payload to JSON (minimal processing)
            String content = payload.toString(); // Faster than ObjectMapper
            long timestamp = System.currentTimeMillis();
            
            // Ultra-fast database insert (prepared statement equivalent)
            jdbc.update("INSERT INTO iot_payload_ultra (id, content, ts) VALUES (?, ?, ?)", 
                       id, content, timestamp);
            
            // No background processing - just return immediately (like NestJS ultra-fast)
            long elapsedNanos = System.nanoTime() - start;
            double elapsedMs = elapsedNanos / 1_000_000.0;
            
            // Reuse object to avoid allocation (like NestJS version)
            responseTemplate.put("id", id);
            responseTemplate.put("t_ms", Math.round(elapsedMs * 100.0) / 100.0);
            
            return ResponseEntity.ok(responseTemplate);
            
        } catch (Exception e) {
            Map<String, Object> error = Map.of("error", "Internal server error");
            return ResponseEntity.status(500).body(error);
        }
    }

    @GetMapping("health")
    public ResponseEntity<Map<String,Object>> health() {
        // Reuse object to avoid allocation (like NestJS version)
        healthTemplate.put("timestamp", new Date().toString());
        healthTemplate.put("count", counter.get());
        
        return ResponseEntity.ok(healthTemplate);
    }
}
